### ISSUE TITLE
Inconsistent evaluation results across runs

### ISSUE THREAD

**Issue by zihuanqiu** [sentient-agi/EvoSkill#6]
Hi, thanks for the great work!

I found that even with the same model, data, and config, the evaluation results vary across runs (sometimes quite noticeably).
This happens even before any evolution/training — just evaluating the base model on the validation set multiple times already gives different results.
Is this expected?
Are there sources of randomness during evaluation?
Is there a way to make the results deterministic?

Not sure if I’m doing something wrong or if this is normal behavior.

Thanks!

**Comment by salzubi401** [comment]
Hi @zihuanqiu, what benchmarks/models are you running on? There's naturally gonna be a source of stochasticity as we're using agents in general. From what I've experimented with for Opus on OfficeQA most runs have been within +-1% of reported.

**Comment by zihuanqiu** [comment]
Hi, thanks for the reply!

I’m using **Kimi 2.5**, and I haven’t done any skill evolution or training — just evaluating the base model.

Specifically, I ran `run_eval` twice independently on the same **205 samples** (OfficeQA). Even in this setting, the results differ noticeably between runs.

Here are the two runs for reference:

**Run 1:**

* tolerance=0.0: 0.5024 (50.24%)
* tolerance=0.001: 0.5317 (53.17%)
* tolerance=0.01: 0.6000 (60.00%)
* tolerance=0.025: 0.6244 (62.44%)
* tolerance=0.05: 0.6585 (65.85%)
* tolerance=0.1: 0.7220 (72.20%)
* avg score: **0.6065**
* exact match: **103 / 205**

**Run 2:**

* tolerance=0.0: 0.4293 (42.93%)
* tolerance=0.001: 0.4634 (46.34%)
* tolerance=0.01: 0.5463 (54.63%)
* tolerance=0.025: 0.6146 (61.46%)
* tolerance=0.05: 0.6537 (65.37%)
* tolerance=0.1: 0.6976 (69.76%)
* avg score: **0.5675**
* exact match: **88 / 205**

So the gap is noticeably large.

Is this still expected for this setup, or could there be something I’m missing?

**Comment by dastin-sentient** [comment]
Thanks for sharing the numbers! 

Short answer: this is expected due to the non-determinism nature of LLM inference and rounding errors of computation. 

In our internal benchmark on OfficeQA on Claude models, we observed the same behavior. The differences between agent's answer and ground truth mostly come from 2 sources: 1) difference between the number shown in raw pdf file and the parsed version of the same file (treasury_bulletin_pdfs) 2) rounding error while doing computation (for example, calculating geometric mean using bash/python). The accuracy is noticeably sensitive while using a low tolerance, and it stabilizes across runs as we increase the tolerance.
